src/Servers/IIS/IIS/src/Core/IISHttpContext.cs

```diff
@@ -385,8 +385,12 @@ public unsafe void SetResponseHeaders()
                             var headerNameBytes = Encoding.UTF8.GetBytes(headerPair.Key);
                             fixed (byte* pHeaderName = headerNameBytes)
                             {
-                                NativeMethods.HttpResponseSetUnknownHeader(_pInProcessHandler, pHeaderName, pHeaderValue, (ushort)headerValueBytes.Length, fReplace: isFirst);
-                            }
+								if(headerValueBytes.Length == 0){
+									throw new ArgumentException("Header value cannot be an empty string");
```

Comment:
`ArgumentException` may not be the most appropriate here (I know I suggested it but I missed the full context 😄). Thoughts @Tratcher @jkotalik ?

Also, in the bug it was mentioned that having the header name be present in the message would be helpful.

Reply:
HttpResponseSetUnknownHeader was already throwing an ArgumentException in this case, right? So is this basically just changing the exception message to make it clearer?

Kestrel throws InvalidOperationExceptions for header names and values that it rejects (though it *does* allow empty header values), so it's probably better to use an InvalidOperationException here too for consistency.

I'm hoping this exception gets bubbled up through the response header dictionary's indexer setter and Add methods, but seeing that SetResponseHeaders is called by ProduceStart which is delayed until the headers are flushed, I'm guessing that's not the case.

We should ensure we throw an exception as early as possible for invalid empty header values and we should add a unit test for it similar to what we [have in Kestrel](https://github.com/aspnet/AspNetCore/blob/a79bb2a79d76a39d31809b71335e5a20e0747ff3/src/Servers/Kestrel/Core/test/HttpResponseHeadersTests.cs#L100).

Reply:
This isn't worth throwing for, just skip the entry. I think you need to check for null or empty before encoding on line 380.